File: src/services/instructions/incdec/INC_W_REG.java
package emulator.source.incdec

import emulator.engine.CpuContext
 import Instruction from '../Instruction'import REGISTER_VALUE_NAME_MAPPER from '@/constants/registers'

export default class INC_W_REG extends Instruction {
	public INC_W_REG(memory, address, source, 
 destination) {
		super(memory, address, source, destination)
		super.setAssembler("inc.w " + this.sdest)
	}
	
	
	exec ({ context, memory }) {
		int old = context.getReg(this.destination) 
		long res = old + 1
		context.getReg(this.destination)  = (int)res
		markFlags(res, context.getReg(this.destination) , context)
		markOverflow(old, 1, (int)res, context)

		context.pc  += 2
	}
}

File: src/services/instructions/incdec/IncWReg.js
import Instruction from '../Instruction'
import { REGISTER_VALUE_NAME_MAPPER } from '@/constants/registers'

const ASSEMBLER_INSTRUCTION_EXPRESSION = (sdestination) => `inc.w  ${sdestination}`

export default class IncWReg extends Instruction {
	constructor (memory, address, source, destination, symbolTable) {
		super(memory, address, source, destination, symbolTable)
		super.setAssembler(ASSEMBLER_INSTRUCTION_EXPRESSION(this.sdestination))
	}

	exec ({ context, memory }) {
		const old = context[REGISTER_VALUE_NAME_MAPPER[this.destination]]
		// TODO: long type -->> long result = old + 1

		const result = old + 1
		//TODO: cast to int type -->>  context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  = (int)result

		context[REGISTER_VALUE_NAME_MAPPER[this.destination]] = result
		Instruction.markFlags(result, context[REGISTER_VALUE_NAME_MAPPER[this.destination]] , context)
		Instruction.markOverflow(old, 1, result, context)

		context.pc  += 2
	}
}
